File: src/Domain/Service/TaskService.php
<?php

namespace App\Domain\Service;

use App\Domain\Dto\TaskCollectionDto;
use App\Domain\Dto\TaskDto;
use App\Domain\Exception\InvalidTaskException;
use App\Domain\Exception\TaskNotFoundException;
use App\Domain\Repository\TaskRepositoryInterface;
use App\Domain\Transformer\TaskCollectionDtoToCollectionTransformer;
use App\Domain\Transformer\TaskDtoToEntityTransformer;
use App\Domain\Validator\TaskValidator;

class TaskService
{
    /**
     * @var TaskRepositoryInterface
     */
    private $repository;
    /**
     * @var TaskValidator
     */
    private $validator;
    /**
     * @var TaskDtoToEntityTransformer
     */
    private $dtoTransformer;
    /**
     * @var TaskCollectionDtoToCollectionTransformer
     */
    private $dtoCollectionTransformer;

    public function __construct(
        TaskRepositoryInterface $repository,
        TaskValidator $validator,
        TaskDtoToEntityTransformer $dtoTransformer,
        TaskCollectionDtoToCollectionTransformer $dtoCollectionTransformer
    ) {
        $this->repository = $repository;
        $this->validator = $validator;
        $this->dtoTransformer = $dtoTransformer;
        $this->dtoCollectionTransformer = $dtoCollectionTransformer;
    }

    /**
     * @throws InvalidTaskException
     */
    public function addTask(TaskDto $taskDto): TaskDto
    {
        $this->validator->validate($taskDto);
        $task = $this->repository->saveTask($this->dtoTransformer->transform($taskDto));

        return $this->dtoTransformer->reverseTransform($task);
    }

    public function getAllTasks(): TaskCollectionDto
    {
        return $this->dtoCollectionTransformer->reverseTransform($this->repository->getTasks());
    }

    public function completeTask(int $taskId): TaskDto
    {
        $task = $this->repository->findTaskById($taskId);
        if (null === $task) {
            throw new TaskNotFoundException("Task not found!");
        }
        $task->complete();
        $this->repository->saveTask($task);

        return $this->dtoTransformer->reverseTransform($task);
    }

    /**
     * @throws InvalidTaskException
     * @throws TaskNotFoundException
     */
    public function updateTask(TaskDto $taskDto) : TaskDto
    {
        $this->validator->validate($taskDto, true);
        $task = $this->repository->findTaskById($taskDto->id);
        if (null === $task) {
            throw new TaskNotFoundException("Task not found!");
        }
        $task = $this->dtoTransformer->transform($taskDto, $task);
        $this->repository->saveTask($task);

        return $this->dtoTransformer->reverseTransform($task);
    }

    public function getTask(int $taskId): TaskDto
    {
        $task = $this->repository->findTaskById($taskId);
        if (null === $task) {
            throw new TaskNotFoundException("Task not found!");
        }

        return $this->dtoTransformer->reverseTransform($task);
    }
}

File: go-api/internal/service/task_service.go
package service

import (
	"fmt"
	"time"

	"github.com/workspace/go-api/internal/model"
	"github.com/workspace/go-api/internal/repository"
)

// TaskNotFoundError is returned when a task cannot be found by ID.
// Error message matches PHP's TaskNotFoundException exactly.
type TaskNotFoundError struct{}

func (e *TaskNotFoundError) Error() string {
	return "Task not found!"
}

// CouldNotDeleteError is returned when a task deletion fails.
// Error message wraps the underlying error, matching PHP's CouldNotDeleteException.
type CouldNotDeleteError struct {
	Msg string
}

func (e *CouldNotDeleteError) Error() string {
	return e.Msg
}

// TaskService contains the business logic for task operations.
type TaskService struct {
	repo *repository.TaskRepository
}

// NewTaskService creates a new TaskService.
func NewTaskService(repo *repository.TaskRepository) *TaskService {
	return &TaskService{repo: repo}
}

// AddTask validates and creates a new task.
func (s *TaskService) AddTask(name, description, when string) (*model.Task, error) {
	if err := ValidateTask(name, description, when, false, nil); err != nil {
		return nil, err
	}

	parsedWhen, _ := time.Parse("2006-01-02", when)
	task := &model.Task{
		Name:        name,
		Description: description,
		When:        parsedWhen,
		Done:        false,
	}

	return s.repo.SaveTask(task)
}

// GetAllTasks retrieves all tasks, optionally filtered by completion status and due date.
func (s *TaskService) GetAllTasks(areDone *bool, when *time.Time) ([]model.Task, error) {
	return s.repo.GetTasks(areDone, when)
}

// GetTask retrieves a single task by ID.
func (s *TaskService) GetTask(taskID int) (*model.Task, error) {
	task, err := s.repo.FindTaskByID(taskID)
	if err != nil {
		return nil, err
	}
	if task == nil {
		return nil, &TaskNotFoundError{}
	}
	return task, nil
}

// CompleteTask marks a task as done.
func (s *TaskService) CompleteTask(taskID int) (*model.Task, error) {
	task, err := s.repo.FindTaskByID(taskID)
	if err != nil {
		return nil, err
	}
	if task == nil {
		return nil, &TaskNotFoundError{}
	}

	task.Done = true
	return s.repo.SaveTask(task)
}

// UpdateTask validates and updates an existing task.
func (s *TaskService) UpdateTask(id int, name, description, when string, rawID interface{}) (*model.Task, error) {
	if err := ValidateTask(name, description, when, true, rawID); err != nil {
		return nil, err
	}

	task, err := s.repo.FindTaskByID(id)
	if err != nil {
		return nil, err
	}
	if task == nil {
		return nil, &TaskNotFoundError{}
	}

	parsedWhen, _ := time.Parse("2006-01-02", when)
	task.Name = name
	task.Description = description
	task.When = parsedWhen

	return s.repo.SaveTask(task)
}

// DeleteTask removes a task by ID, returning the task data before deletion.
func (s *TaskService) DeleteTask(taskID int) (*model.Task, error) {
	task, err := s.repo.FindTaskByID(taskID)
	if err != nil {
		return nil, err
	}
	if task == nil {
		return nil, &TaskNotFoundError{}
	}

	// Copy task data before deletion (matches PHP behavior)
	taskCopy := *task

	if err := s.repo.DeleteTask(task); err != nil {
		return nil, &CouldNotDeleteError{Msg: fmt.Sprintf("Could not delete the task: %s", err.Error())}
	}

	return &taskCopy, nil
}
